In [34]:
#Pandas: Panel Data System.
#Collection of features which belong to a certin domain which is arranged in layers or groups.
#Data Series: Single column data. (Time Series Analysis)
#Data Frames: More than one column of data.
#List 1D
#Tuple Immutable, 1D
#Set Not ordered

import pandas as pd
data = [['Srishti', 1, 'A'],['Goutham',2,'B'],['Sanskaar',3,'C'],['Julian',4,'O',95]] #list and tuple we can't use column headers
print(type(data))
print(data)
df = pd.DataFrame(data)
print(df)


<class 'list'>
[['Srishti', 1, 'A'], ['Goutham', 2, 'B'], ['Sanskaar', 3, 'C'], ['Julian', 4, 'O', 95]]
          0  1  2     3
0   Srishti  1  A   NaN
1   Goutham  2  B   NaN
2  Sanskaar  3  C   NaN
3    Julian  4  O  95.0


In [35]:
data_t = (('Srishti', 1, 'A'),('Goutham',2,'B'),('Sanskaar',3,'C'),('Julian',4,'O',95))#list and tuple we can't use column headers
print(type(data_t))
print(data_t)
df = pd.DataFrame(data_t)
print(df)

<class 'tuple'>
(('Srishti', 1, 'A'), ('Goutham', 2, 'B'), ('Sanskaar', 3, 'C'), ('Julian', 4, 'O', 95))
          0  1  2     3
0   Srishti  1  A   NaN
1   Goutham  2  B   NaN
2  Sanskaar  3  C   NaN
3    Julian  4  O  95.0


In [36]:
# Representation of a dictionary in a column format where keys are the column headers and values are the rows
data_d = {'Name':['Srishti', 'Goutham', 'Sankar'], 'RegisterNumber': [1,2,3]}
print(type(data_d))
print(data_d)
df_d = pd.DataFrame(data_d)
print(df_d)


<class 'dict'>
{'Name': ['Srishti', 'Goutham', 'Sankar'], 'RegisterNumber': [1, 2, 3]}
      Name  RegisterNumber
0  Srishti               1
1  Goutham               2
2   Sankar               3


In [37]:
import numpy as np 
d = [[1,2,3],[4,5,6],[7,8,9]]
d1 = np.array(d)
print(type(d1))
df = pd.DataFrame(d1)
print(df)

<class 'numpy.ndarray'>
   0  1  2
0  1  2  3
1  4  5  6
2  7  8  9


In [38]:
print(df_d)
print(df_d['RegisterNumber'])
print(type(df_d['RegisterNumber']))

      Name  RegisterNumber
0  Srishti               1
1  Goutham               2
2   Sankar               3
0    1
1    2
2    3
Name: RegisterNumber, dtype: int64
<class 'pandas.core.series.Series'>


In [39]:
# data = np.random.rand(2,2)
# print(type(data))
df = pd.DataFrame(data)
df = pd.DataFrame(np.random.rand(2,2), columns = ['c1','c2'], index = ['r1','r2'])
print(df)


          c1        c2
r1  0.700414  0.571568
r2  0.640997  0.641006


# Find the output for the following cases in dataframe creation
1. List of dictionaries
2. Dictionary of Lists.
3. Dictionaries of Dictionaries.
4. Data Series appended column wise.(Concatenation)

In [40]:
dict_list = [{'Name':'Srishti', 'RegisterNumber': 1}, {'Name':'Goutham', 'RegisterNumber': 2},{'Name':'Daniel', 'RegisterNumber': 3}]
print(type(dict_list))
print(dict_list)
df_d = pd.DataFrame(dict_list)
print(df_d)

<class 'list'>
[{'Name': 'Srishti', 'RegisterNumber': 1}, {'Name': 'Goutham', 'RegisterNumber': 2}, {'Name': 'Daniel', 'RegisterNumber': 3}]
      Name  RegisterNumber
0  Srishti               1
1  Goutham               2
2   Daniel               3


In [41]:
dictionary_of_lists = {
    "fruits": ["apple", "banana", "cherry"],
    "colors": ["red", "yellow", "pink"],
    "numbers": [1, 2, 3]
}
print(type(dictionary_of_lists))
print(dictionary_of_lists)
df_d = pd.DataFrame(dictionary_of_lists)
print(df_d)

<class 'dict'>
{'fruits': ['apple', 'banana', 'cherry'], 'colors': ['red', 'yellow', 'pink'], 'numbers': [1, 2, 3]}
   fruits  colors  numbers
0   apple     red        1
1  banana  yellow        2
2  cherry    pink        3


In [42]:
dictionary_of_dictionaries = {
    "person1": {
        "name": "Alice",
        "age": 30,
        "city": "New York"
    },
    "person2": {
        "name": "Bob",
        "age": 25,
        "city": "Los Angeles"
    },
    "person3": {
        "name": "Charlie",
        "age": 35,
        "city": "Chicago"
    }
}

print(type(dictionary_of_dictionaries))
# print(dictionary_of_dictionaries)
df_d = pd.DataFrame(dictionary_of_dictionaries)
print(df_d)


<class 'dict'>
       person1      person2  person3
name     Alice          Bob  Charlie
age         30           25       35
city  New York  Los Angeles  Chicago


In [43]:
import pandas as pd

list1 = [1, 2, 3, 4]
list2 = [5, 6, 7, 8]

df = pd.DataFrame({'Column1': list1, 'Column2': list2})

print(df)



   Column1  Column2
0        1        5
1        2        6
2        3        7
3        4        8


In [44]:
dict1 = {'Name': 'Dijin','Register Number': 1}
dict2 = {'Name': 'Joseph','Register Number': 2}

df = pd.DataFrame([dict1,dict2])

print(df)

     Name  Register Number
0   Dijin                1
1  Joseph                2


## September 15

In [1]:
Name = ["Dijin", "Sagar"]
Age = [22,21]
x = (Name, Age)
print("Before Zipping:\n    ", x)
x = zip(Name, Age)
print("After Zipping:   ",x)
print(type(x))
print(x)

Before Zipping:
     (['Dijin', 'Sagar'], [22, 21])
After Zipping:    <zip object at 0x000002987EA89E00>
<class 'zip'>


In [46]:
# data = list(x)
import pandas as pd 
df = pd.DataFrame(x)
df

0   1
0  Dijin  22
1  Sagar  21

In [47]:
data = {'col1':pd.Series([1,2,3], index = ['r1','r2','r3']), 'col2':pd.Series([10,20,30], index = ['r1','r2','r3'])}
df = pd.DataFrame(data)
df

col1  col2
r1     1    10
r2     2    20
r3     3    30

In [48]:
data1 = pd.read_csv("data.csv")
data1

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0    ...          17.33           184.60      2019.0           0.16220   
1    ...          23.41           158.80      1956.0           0.12380   
2    ...          25.53           152.50      1709.0           0.14440   
3    ...          26.50            98.87       567.7           0.20980   
4    ...          16.67           152.20      1575.0           0.13740   
..   ...            ...              ...         ...               ...   
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0              0.66560           0.7119                0.2654          0.4601   
1              0.18660           0.2416                0.1860          0.2750   
2              0.42450           0.4504                0.2430          0.3613   
3              0.86630           0.6869                0.2575          0.6638   
4              0.20500           0.4000                0.1625          0.2364   
..                 ...              ...                   ...             ...   
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
0                    0.11890          NaN  
1                    0.08902          NaN  
2                    0.08758          NaN  
3                    0.17300          NaN  
4      

In [49]:
data1.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [50]:
data1.tail()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
564  926424         M        21.56         22.39          142.00     1479.0   
565  926682         M        20.13         28.25          131.20     1261.0   
566  926954         M        16.60         28.08          108.30      858.1   
567  927241         M        20.60         29.33          140.10     1265.0   
568   92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
564                  0.07115          NaN  
565                  0.06637          NaN  
566                  0.07820          NaN  
567                  0.12400          NaN  
568                  0.07039          NaN  

[5 rows x 33 columns]

In [51]:
data1.describe()

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  texture_worst  perimeter_worst   area_worst  \
count     569.000000  ...     569.000000       569.000000   569.000000   
mean        0.181162  ...      25.677223       107.261213   880.583128   
std         0.027414  ...       6.146258        33.602542   569.356993   
min         0.106000  ...      12.020000        50.410000   185.200000   
25%         0.161900  ...      21.080000        84.110000   515.300000   
50%         0.179200  ...      25.410000        97.660000   686.500000   
75%         0.195700  ...      29.720000       125.400000  1084.000000   
max         0.304000  ...      49.540000       251.200000  4254.000000   

       smoothness_worst  compactness_worst  concavity_worst  \
count        569.000000         569.000000       569.000000   
mean           0.132369           0.254265         0.272188   
std            0.022832           0.157336         0.208624   
min            0.071170           0.027290         0.000000   
25%            0.116600           0.147200         0.114500   
50%            0.131300           0.211900         0.226700   
75%            0.146000           0.339100         0.382900   
max            0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  \
count            569.000000      569.000000               569.000000   
mean               0.114606        0.290076                 0.083946   
std                0.065732        0.061867                 0.018061   
min                0.000000        0.156500                 0.055040   
25%                0.064930        0.250400                 0.071460   
50%                0.099930        0.282200                 0.080040   
75%                0.161400        0.317900                 0.092080   
max                0.291000        0.663800                 0.207500   

       Unnamed: 32  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 32 columns]

In [52]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [53]:
import numpy as np
import pandas as pd

XYZ_mall = {'Days1': [1,2,3,4,5,6,7], "Sales1" : [1000,1500,3000,1000,4000,5000,4000]}
XYZ_road = {'Days2' : [1,2,3,4,5,6,7], "Sales2" : [10000,4500,6000,5500,4300,6000,3000]}
df1 = pd.DataFrame(XYZ_mall)
df2 = pd.DataFrame(XYZ_road)
print("Pizza Hut\n", df1)
print("99VarietyDosa\n", df2)

Pizza Hut
    Days1  Sales1
0      1    1000
1      2    1500
2      3    3000
3      4    1000
4      5    4000
5      6    5000
6      7    4000
99VarietyDosa
    Days2  Sales2
0      1   10000
1      2    4500
2      3    6000
3      4    5500
4      5    4300
5      6    6000
6      7    3000


In [54]:
print(df1.describe())
print(df2.describe())

          Days1       Sales1
count  7.000000     7.000000
mean   4.000000  2785.714286
std    2.160247  1629.344011
min    1.000000  1000.000000
25%    2.500000  1250.000000
50%    4.000000  3000.000000
75%    5.500000  4000.000000
max    7.000000  5000.000000
          Days2        Sales2
count  7.000000      7.000000
mean   4.000000   5614.285714
std    2.160247   2211.657426
min    1.000000   3000.000000
25%    2.500000   4400.000000
50%    4.000000   5500.000000
75%    5.500000   6000.000000
max    7.000000  10000.000000


In [55]:
print("Average Sales for Pizza Hut:",df1['Sales1'].mean())
print("Average Sales for 99VarietyDosa:",df2['Sales2'].mean())


Average Sales for Pizza Hut: 2785.714285714286
Average Sales for 99VarietyDosa: 5614.285714285715


In [56]:
print("Total Sales for Pizza Hut:",df1['Sales1'].sum())
print("Total Sales for 99VarietyDosa:",df2['Sales2'].sum())

Total Sales for Pizza Hut: 19500
Total Sales for 99VarietyDosa: 39300


In [57]:

df = pd.DataFrame({'Column1': list1, 'Column2': list2})

print(df)

   Column1  Column2
0        1        5
1        2        6
2        3        7
3        4        8


1. Find the total sales made by the company Marico

2. Find the highest and the least sales of the company along with its Quater

3. Find out the quaters which are earning more than 130 crs.

4. Find the difference in half yearly performance of the company

5. If the total sales of Marico in the financial year 2022-23 was 180 billions find out the difference in this years performance.

6. If the expected sales for the next year is 600 crores, find out the difference in this years and next years performance and create a new daaframe to store the details.

7. Find out the average performance of the company for the financial year 22-23. Check out the deviation from the past year if the average sales is found to be 45 billion dollars.



In [58]:
import pandas as pd
df = pd.DataFrame({'Quaters': ['Q1', 'Q2','Q3','Q4'], 'Sales(inCrores)': [100,120,90,150] })

print(df)

  Quaters  Sales(inCrores)
0      Q1              100
1      Q2              120
2      Q3               90
3      Q4              150


In [59]:
print("The total sales by the company is", df['Sales(inCrores)'].sum(), "Crores.")

The total sales by the company is 460 Crores.


In [60]:
print("The least sales of Marico is in and it's", df['Sales(inCrores)'].min())

The least sales of Marico is in and it's 90


In [61]:
print("The least sales of Marico is in and it's", df['Sales(inCrores)'].max())
print(df['Sales(inCrores)'].max().df['Quaters']

SyntaxError: incomplete input (584905480.py, line 2)

# SEPTEMBER 19


In [112]:
import pandas as pd
data = [110001, 800001, 560029,600001]
df = pd.DataFrame(data, columns = ["Pincode"])
print(df)
Place = ['Delhi', 'Patna', 'Bangalore', 'Chennai']
Place_Type = ['Urban', 'Suburb', 'Urban', 'Suburb']
df['Place'] = Place 
df['Place_Type'] = Place_Type
df.rename(columns = {'Pincode' : 'PINCODE'}, inplace = True)
print(df)

   Pincode
0   110001
1   800001
2   560029
3   600001
   PINCODE      Place Place_Type
0   110001      Delhi      Urban
1   800001      Patna     Suburb
2   560029  Bangalore      Urban
3   600001    Chennai     Suburb


In [ ]:
df = df.sort_values(by = 'Place')
df

In [ ]:
df = df.sort_values (by= ["PINCODE", "Place_Type"])
df

PINCODE      Place Place_Type
0   110001      Delhi      Urban
2   560029  Bangalore      Urban
3   600001    Chennai     Suburb
1   800001      Patna     Suburb

In [ ]:
df = df.sort_index
df

<bound method DataFrame.sort_index of    PINCODE      Place Place_Type
0   110001      Delhi      Urban
2   560029  Bangalore      Urban
3   600001    Chennai     Suburb
1   800001      Patna     Suburb>

#### Extract Pincode from Dataframe df and convert it into data series
#### Add a new row with the following values.(577501, "Chitraturga"), "Suburb")

In [ ]:
se  = df['PINCODE']
print(type(se))
print(se)

<class 'pandas.core.series.Series'>
0    110001
1    800001
2    560029
3    600001
Name: PINCODE, dtype: int64


In [113]:
n = pd.DataFrame([577501, "Chitraturga", "Suburb"])
n

0
0       577501
1  Chitraturga
2       Suburb

In [68]:
new_df = df.append(n, ignore_index = True, axis = 1)
new_df

AttributeError: 'DataFrame' object has no attribute 'append'

In [114]:
new_df = pd.concat([df,n])
new_df

PINCODE      Place Place_Type            0
0  110001.0      Delhi      Urban          NaN
1  800001.0      Patna     Suburb          NaN
2  560029.0  Bangalore      Urban          NaN
3  600001.0    Chennai     Suburb          NaN
0       NaN        NaN        NaN       577501
1       NaN        NaN        NaN  Chitraturga
2       NaN        NaN        NaN       Suburb

In [94]:
data = pd.DataFrame({'a': [1,2], 'b': [3,4], 'c': [5,6], 'd': "M"}, index_col = "d")
data

TypeError: DataFrame.__init__() got an unexpected keyword argument 'index_col'

In [86]:
data.iloc[0,0]

1

In [87]:
data.iloc[0]

a    1
b    3
c    5
d    M
Name: 0, dtype: object

In [93]:
data.loc["d"]

KeyError: 'd'

Can loc work on multipl columns, Explore.

In [98]:
print(df.isnull())
print(new_df.isnull())

   PINCODE  Place  Place_Type
0    False  False       False
1    False  False       False
2    False  False       False
3    False  False       False
   PINCODE  Place  Place_Type      0
0    False  False       False   True
1    False  False       False   True
2    False  False       False   True
3    False  False       False   True
0     True   True        True  False
1     True   True        True  False
2     True   True        True  False


In [99]:
print(df.notnull())
print(new_df.notnull())

   PINCODE  Place  Place_Type
0     True   True        True
1     True   True        True
2     True   True        True
3     True   True        True
   PINCODE  Place  Place_Type      0
0     True   True        True  False
1     True   True        True  False
2     True   True        True  False
3     True   True        True  False
0    False  False       False   True
1    False  False       False   True
2    False  False       False   True


In [101]:
print(new_df.dropna())

Empty DataFrame
Columns: [PINCODE, Place, Place_Type, 0]
Index: []


In [ ]:
# import sklearn
# from sklearn.impute import SimpleImputer
# imputer = SimpleImputer(Missing)

In [103]:
new_df = new_df.fillna("a")
new_df

PINCODE      Place Place_Type            0
0  110001.0      Delhi      Urban            a
1  800001.0      Patna     Suburb            a
2  560029.0  Bangalore      Urban            a
3  600001.0    Chennai     Suburb            a
0         a          a          a       577501
1         a          a          a  Chitraturga
2         a          a          a       Suburb

### If in the given dataframe odd rows are having missing values fill it with one even rows fill it with zero

df.loc

In [105]:
df.loc()

In [107]:
df.loc[1]

PINCODE       800001
Place          Patna
Place_Type    Suburb
Name: 1, dtype: object

Explore concat and merge with respect to the dataframes

In [134]:
import pandas as pd
data = [110001, 800001, 560029,600001]
df = pd.DataFrame(data, columns = ["Pincode"])
Place = ['Delhi', 'Patna', 'Bangalore', 'Chennai']
Place_Type = ['Urban', 'Suburb', 'Urban', 'Suburb']
df['Place'] = Place 
df['Place_Type'] = Place_Type
df.rename(columns = {'Pincode' : 'PINCODE'}, inplace = True)
print(df)
n = pd.DataFrame([577501, "Chitraturga", "Suburb"])
new_df = pd.concat([df,n], ignore_index=True)
print(new_df)

for i in new_df.index:
    i = int(i)
    if i+ 1 % 2 == 0:
        new_df = new_df.loc[i].fillna(0)
    else:
        new_df = new_df.loc[i].fillna(1)
new_df

   PINCODE      Place Place_Type
0   110001      Delhi      Urban
1   800001      Patna     Suburb
2   560029  Bangalore      Urban
3   600001    Chennai     Suburb
    PINCODE      Place Place_Type            0
0  110001.0      Delhi      Urban          NaN
1  800001.0      Patna     Suburb          NaN
2  560029.0  Bangalore      Urban          NaN
3  600001.0    Chennai     Suburb          NaN
4       NaN        NaN        NaN       577501
5       NaN        NaN        NaN  Chitraturga
6       NaN        NaN        NaN       Suburb


KeyError: 1

In [135]:
new_df.index

Index(['PINCODE', 'Place', 'Place_Type', 0], dtype='object')

In [117]:
for i in new_df.index:
    print(i)

0
1
2
3
0
1
2


In [119]:
new_df.iloc[1].fillna(0)

PINCODE       800001.0
Place            Patna
Place_Type      Suburb
0                    0
Name: 1, dtype: object

In [126]:
data = {'A': [1, 2, np.nan, 4],
        'B': [np.nan, 2, 3, 4],
        'C': [1, np.nan, np.nan, 4]}

df1 = pd.DataFrame(data)
df1

A    B    C
0  1.0  NaN  1.0
1  2.0  2.0  NaN
2  NaN  3.0  NaN
3  4.0  4.0  4.0

In [131]:
for i in df1.index:
    if (i + 1) % 2 == 0:
        df1 = df1.iloc[i].fillna(0)
    else:
        df1 = df1.iloc[i].fillna(1)
df1

TypeError: can only concatenate str (not "int") to str

In [132]:
df1.index

Index(['A', 'B', 'C'], dtype='object')

In [136]:
df1 = pd.DataFrame({"a":[100,200,300,400], "b":[40,32,23,64]})
df1

df2 = pd.DataFrame({"c":[91,22,63], "d":[11,52,73], "e":[13,27,34]})
df2

df3 = pd.concat([df1,df2], ignore_index = True)

for i in range(len(df3)):
    if i%2 == 0:
        df3.loc[i].fillna(1, inplace =True)
    else:
        df3.loc[i].fillna(0, inplace =True)

df3

C:\Users\DIJIN\AppData\Local\Temp\ipykernel_24800\2324415287.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3.loc[i].fillna(1, inplace =True)
C:\Users\DIJIN\AppData\Local\Temp\ipykernel_24800\2324415287.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3.loc[i].fillna(0, inplace =True)


a     b     c     d     e
0  100.0  40.0   NaN   NaN   NaN
1  200.0  32.0   NaN   NaN   NaN
2  300.0  23.0   NaN   NaN   NaN
3  400.0  64.0   NaN   NaN   NaN
4    NaN   NaN  91.0  11.0  13.0
5    NaN   NaN  22.0  52.0  27.0
6    NaN   NaN  63.0  73.0  34.0

In [137]:
df1 = pd.DataFrame({"a":[100,200,300,400], "b":[40,32,23,64]})
df1

a   b
0  100  40
1  200  32
2  300  23
3  400  64

In [138]:
df2 = pd.DataFrame({"c":[91,22,63], "d":[11,52,73], "e":[13,27,34]})
df2

c   d   e
0  91  11  13
1  22  52  27
2  63  73  34

In [139]:
df3 = pd.concat([df1,df2], ignore_index = True)
df3

a     b     c     d     e
0  100.0  40.0   NaN   NaN   NaN
1  200.0  32.0   NaN   NaN   NaN
2  300.0  23.0   NaN   NaN   NaN
3  400.0  64.0   NaN   NaN   NaN
4    NaN   NaN  91.0  11.0  13.0
5    NaN   NaN  22.0  52.0  27.0
6    NaN   NaN  63.0  73.0  34.0

In [144]:
df1 = pd.DataFrame({"a":[100,200,300,400], "b":[40,32,23,64]})
df1

df2 = pd.DataFrame({"c":[91,22,63], "d":[11,52,73], "e":[13,27,34]})
df2

df3 = pd.concat([df1,df2], ignore_index = True)
df3

a     b     c     d     e
0  100.0  40.0   NaN   NaN   NaN
1  200.0  32.0   NaN   NaN   NaN
2  300.0  23.0   NaN   NaN   NaN
3  400.0  64.0   NaN   NaN   NaN
4    NaN   NaN  91.0  11.0  13.0
5    NaN   NaN  22.0  52.0  27.0
6    NaN   NaN  63.0  73.0  34.0

In [146]:
for i in range(len(df3)):
    if i%2 == 0:
        df3 = df3.loc[i].fillna(1)
    else:
        df3 = df3.loc[i].fillna(0)

df3

KeyError: 1

In [10]:
import re
st = "This is a sample string with multiple string 12wer 34  "

z = re.finditer("\b\d+\w+", st)

for i in z:
    print(i)